File: Python/KNN.py
import csv
import sys
import math
import numpy as np
from collections import Counter

def KNN(trainDirectory,testdirectory):

    with open(trainDirectory) as train, open(testdirectory) as test:
        reader = csv.reader(train,csv.QUOTE_NONNUMERIC)
        next(reader) # skip the labels line
        memory = [np.array([item for item in row],dtype='i') for row in reader]
        train.close()

        # this choice has roughly as many effective parameters as image dimensions * categories
        k = len(memory)/(10*len(memory[0]));
        k = k - (k % 2) + 1;

        reader = csv.reader(test,csv.QUOTE_NONNUMERIC)
        next(reader) # skip the labels line
        probes = [np.array([item for item in row],dtype='i') for row in reader]
        test.close()

        runKNN(memory,k,probes)


def runKNN(memory,k,probes):

    with open('answers', 'w') as answerFile:
        answerFile.write("ImageId,Label")
        for toProbe in xrange(len(probes)):
            answer = knnVote(memory,k,probes[toProbe]);
            answerFile.write(str(toProbe+1) + ",\"" + str(answer) + '\"\n')
            print str(answer) + '\t' + str(float(1+toProbe)/len(probes))
        answerFile.flush()
        answerFile.close()


def knnVote(memory,k,probe):

    q = MaxHeap(k)
    for row in xrange(len(memory)):
        distance = np.linalg.norm(probe - memory[row][1:])
        q.pushPop([distance,memory[row][0]])
    return Counter([item[1] for item in q.data]).most_common(1)[0][0]


class MaxHeap:
    def __init__(self,k):
        self.data = [[sys.maxint, 15] for i in xrange(k)]
    def pushPop(self,entry):
        if (entry[0] < self.data[0][0]):
            self.addToHeap(entry,1)
    def addToHeap(self,entry,vertex):
        if 2*vertex-1 < len(self.data) and self.data[2*vertex-1][0] > entry[0]:
            self.data[vertex-1] = self.data[2*vertex-1]
            self.addToHeap(entry,2*vertex)
        elif 2*vertex < len(self.data) and self.data[2*vertex][0] > entry[0]:
            self.data[vertex-1] = self.data[2*vertex]
            self.addToHeap(entry,2*vertex+1)
        else:
            self.data[vertex-1] = entry

KNN("../Data/train.csv","../Data/test.csv")

File: R/KNN.r
KNN <- function(trainDirectory,testdirectory) {

	memory <- read.csv(trainDirectory)

	# this choice has roughly as many effective parameters as image dimensions * categories
	k <- nrow(memory)/(10*ncol(memory))
	k <- k - (k %% 2) + 1;

	probes <- read.csv(testdirectory)

	runKNN(memory,k,probes)
}

runKNN <- function(memory,k,probes) {

	df = data.frame(matrix(ncol = 2, nrow = nrow(probes)))
	for (toProbe in 1:nrow(probes)) {
		answer <- knnVote(memory,k,as.integer(probes[toProbe,]))
		df[toProbe,] <- c(toProbe,answer)
		print(paste(toString(answer),toString(toProbe/nrow(probes)),sep="\t"))
	}
	colnames(df) <- c("ImageId", "Label")
	write.csv(df,file="answers.csv",row.names=FALSE)
}

knnVote <- function(memory,k,probe) {

	q <- MaxHeap(k)
	for (row in 1:nrow(memory)) {
		memoryLoc <- as.integer(memory[row,2:ncol(memory)])
		distance <- norm(as.matrix(probe-memoryLoc))
		q <- pushPop.MaxHeap(q,c(distance,as.integer(memory[row,1])))
	}
	categories <- as.integer(q$data[,2])
	uniqueCategories <- unique(categories)
	return(uniqueCategories[which.max(tabulate(match(categories,uniqueCategories)))])
}

MaxHeap <- function(k) {

	ma <- matrix(c(.Machine$integer.max,15), nrow=1)
	data <- ma[rep(1,k),]

	me <- list(data=data)

	class(me) <- "MaxHeap"

	return(me)
}

pushPop <- function(x) UseMethod("pushPop")
pushPop.default <- function(x) "Unknown class"

pushPop.MaxHeap <- function(maxHeap,entry) {

			
	if (entry[1] < maxHeap$data[1,1]) {
		maxHeap$data <- addToHeap(maxHeap$data,entry,1)
	}
	
	return(maxHeap)
}

addToHeap <- function(data,entry,vertex) {

	if (2*vertex <= nrow(data) && data[2*vertex,1] > entry[1]) {
		data[vertex,] <- data[2*vertex,]
		addToHeap(data,entry,2*vertex)
	} else if (2*vertex+1 <= nrow(data) && data[2*vertex+1,1] > entry[1]) {
		data[vertex,] <- data[2*vertex+1,]
		addToHeap(data,entry,2*vertex+1)
	} else {
		data[vertex,] <- entry
		return(data)
	}
}


KNN("../Data/train.csv","../Data/test.csv")
